src/Qualify.hs

```diff
@@ -11,147 +27,293 @@ import qualified Set
-                                -- fold recursively -- is there a more efficient way?
-                                XObj (Lst (_ : ys)) _ _ ->
-                                  foldl' folder innerEnv ys
-                                x ->
-                                  error ("Can't match variable with " ++ show x)
-                          r' = setFullyQualifiedSymbols typeEnv globalEnv innerEnv' r
-                       in [l', r']
-                    XObj {} ->
-                      let l' = setFullyQualifiedSymbols typeEnv globalEnv env l
-                          r' = setFullyQualifiedSymbols typeEnv globalEnv env r
-                       in [l', r']
-              )
-              (pairwise casesXObjs)
-       in XObj (Lst (matchExpr : newExpr : concat newCasesXObjs)) i t
-    else XObj (Lst (matchExpr : expr : casesXObjs)) i t -- Leave it untouched for the compiler to find the error.
-setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst [XObj With _ _, XObj (Sym path _) _ _, expression]) _ _) =
+     in foldM qualifyBinding (innerEnv, []) (pairwise bindings)
+          >>= \(innerEnv', qualifiedBindings) ->
+            setFullyQualifiedSymbols typeEnv globalEnv innerEnv' body
+              >>= pure . runQualified
+              >>= \qualifiedBody -> pure $ Qualified $ XObj (Lst [letExpr, XObj (Arr qualifiedBindings) bindi bindt, qualifiedBody]) i t
+  where
+    qualifyBinding :: (Env, [XObj]) -> (XObj, XObj) -> Either QualificationError (Env, [XObj])
+    qualifyBinding (e, bs) (s@(XObj (Sym (SymPath _ binderName) _) _ _), o) =
+      setFullyQualifiedSymbols typeEnv globalEnv e o
+        >>= pure . runQualified
+        >>= \qualified -> pure $ (extendEnv e binderName s, bs ++ [s, qualified])
+    qualifyBinding _ _ = error "bad let binding"
+qualifyLet _ _ _ xobj = Left $ FailedToQualifySymbols xobj
+
+-- | Qualify symbols in a Match form.
+qualifyMatch :: Qualifier
+qualifyMatch typeEnv globalEnv env (XObj (Lst (matchExpr@(XObj (Match _) _ _) : expr : casesXObjs)) i t)
+  | odd (length casesXObjs) = pure $ Qualified $ XObj (Lst (matchExpr : expr : casesXObjs)) i t -- Leave it untouched for the compiler to find the error.
+  | otherwise =
+    setFullyQualifiedSymbols typeEnv globalEnv env expr
+      >>= pure . runQualified
+      >>= \qualifiedExpr ->
+        mapM qualifyCases (pairwise casesXObjs)
+          >>= pure . map (map runQualified)
+          >>= \qualifiedCases -> pure $ Qualified $ XObj (Lst (matchExpr : qualifiedExpr : concat qualifiedCases)) i t
+  where
+    Just ii = i
+    lvl = envFunctionNestingLevel env
+    innerEnv :: Env
+    innerEnv = Env Map.empty (Just env) (Just ("case-env-" ++ show (infoIdentifier ii))) Set.empty InternalEnv lvl
+    qualifyCases :: (XObj, XObj) -> Either QualificationError [Qualified]
+    qualifyCases (l@(XObj (Lst (_ : xs)) _ _), r) =
+      let innerEnv' = foldl' foldVars innerEnv xs
+       in setFullyQualifiedSymbols typeEnv globalEnv env l
```

Comment:
Do notation must look better here, no?

Reply:
(see line 258...)

Reply:
I changed pretty much all these instances of `>>=` to do notation based on our discussions!